GitHubCI: add build/pack/push steps for Xamarin.Forms.nuspec

parhamsaremi opened the pull request:
- Update main.yml file to have build/pack/push steps for Xamarin.Forms.nuspec.
- Update Xamarin.Forms.Platform.GTK to depend on the newly built version of DotNetForms instead of Xamarin.Forms

aarani committed:
[macOS] Fix back button for nested nav pages

In the current implementation of NativeToolbarTracker,
having a nested navigation page causes the toolbar to get
overwritten by the child navigation page which in turns
causes the back button to be unusable because it only moves
in the scope of the child navigation stack so to mitigate
the issue this commit adds the possiblity for the back button
to move out of the child navigation stack back into the parent
navigation stack.

Fixes #8843
Upstream PR: https://github.com/xamarin/Xamarin.Forms/pull/14029

parhamsaremi committed:
GitHubCI: build/pack Xamarin.Forms.nuspec and push

parhamsaremi committed:
GitHubCI: fix NU1101 error

Error message was: D:\a\Xamarin.Forms\Xamarin.Forms\Xamarin.Forms.Build.Tasks\Xamarin.Forms.Build.Tasks.csproj : error NU1101: Unable to find package XliffTasks. No packages exist with this id in source(s): Microsoft Visual Studio Offline Packages, nuget.org [D:\a\Xamarin.Forms\Xamarin.Forms\Xamarin.Forms.sln]

parhamsaremi committed:
GitHubCI: remove UAP v14393 to fix nuget pack

The error message was: Could not find a part of the path 'D:\a\Xamarin.Forms\Xamarin.Forms\Xamarin.Forms.Platform.UAP\bin\Release\uap10.0.14393'.

knocte commented:
> GitHubCI: Remove UAP v14393 to fix nuget pack

Nit: in the first two commits you use lowercase after ':', let's do the same above ^

knocte commented:
3 things that need to be done in this PR (in particular, in the 1st commit of it):
* In the same way it was done in Afshin's commit (https://github.com/nblockchain/DotNetForms/commit/cf6f7a9dcfdb6d765c8612a7bc2ec92e53e32179), change other elements of the nuspec to point to DotNetForms fork.
* Change the file .nuspec/Xamarin.Forms.Platform.GTK.nuspec to depend on DotNetForms (instead of Xamarin.Forms), since thanks to this PR the package now will exist.
* Revert the changes that were done in the file Xamarin.Forms.Platform.GTK/Xamarin.Forms.Platform.GTK.csproj in Afshin's commit, since they shouldn't be needed anymore.

knocte commented:
Oh and one last thing I forgot to mention: please cherry-pick in your branch this commit from Afshin, which is the main reason why we need a DotNetForms package: https://github.com/nblockchain/DotNetForms/pull/2/commits/0d8a8315e6f873d4d21760b9446f1cce9f911bd6

knocte commented:
Great, this is almost ready, but I have 3 comments that need to be addressed first:
* wrt PR title and description: "Update build workflow" doesn't say much info. You can still keep the title short but still say what this PR is mainly doing. The description is based on a Xamarin.Forms template and we don't care about that, just remove all that and explain what the PR does.
* if you're overriding the $env:DOTNETFORMS_VERSION variable, what's the point of it having the initial value 5.0.0.2515? wouldn't it work if you don't even set an initial value?
* I'm worried that the variable $version might get a different value when generating the main package versus when generating the GTK package, shouldn't we use the same value?

parhamsaremi commented:
> Great, this is almost ready, but I have 3 comments that need to be addressed first:
> 
> * wrt PR title and description: "Update build workflow" doesn't say much info. You can still keep the title short but still say what this PR is mainly doing. The description is based on a Xamarin.Forms template and we don't care about that, just remove all that and explain what the PR does.
> * if you're overriding the $env:DOTNETFORMS_VERSION variable, what's the point of it having the initial value 5.0.0.2515? wouldn't it work if you don't even set an initial value?
> * I'm worried that the variable $version might get a different value when generating the main package versus when generating the GTK package, shouldn't we use the same value?

- ok sure I'll change the description and the title.
- I didn't know if yaml supported empty initialization. I'll check that right now.
- you are right I think once I override DOTNETFORMS version, I should use it for both $version and $core_version

knocte commented:
Looking good. If the last push passes CI (green), then there's only one last thing to do here: add a refactoring commit.
What I think should go in this refactoring commit:
- Create an $env:CONFIGURATION=Release variable, and then use that in all places where we use "Release" (even in Afshin's steps), and remove the `$buildConfiguration` var.
- Move both nuget push calls into a single "push" step, to not have to repeat the if conditions `If ($Env:GITHUB_REF -eq $Env:DEFAULT_REF_TO_NUGET_PUSH) {  If ('${{ secrets.NUGET_API_KEY }}' -ne '') ...`.
- Make both nuget push commands use -NoSources (so that you can use * instead of the specific file name).
- Any more simplifications you might think of?

parhamsaremi commented:
> Looking good. If the last push passes CI (green), then there's only one last thing to do here: add a refactoring commit. What I think should go in this refactoring commit:
> 
> * Create an $env:CONFIGURATION=Release variable, and then use that in all places where we use "Release" (even in Afshin's steps), and remove the `$buildConfiguration` var.
> * Move both nuget push calls into a single "push" step, to not have to repeat the if conditions `If ($Env:GITHUB_REF -eq $Env:DEFAULT_REF_TO_NUGET_PUSH) {  If ('${{ secrets.NUGET_API_KEY }}' -ne '') ...`.
> * Make both nuget push commands use -NoSources (so that you can use * instead of the specific file name).
> * Any more simplifications you might think of?

One more thing that I wanted to add: how Afshin is replacing $version$, $core_versoin$, and $configuration$ can be simplified using -p in the pack command :thinking:

knocte commented:
> how Afshin is replacing $version$, $core_versoin$, and $configuration$ can be simplified using -p in the pack command

Ok good idea, let's do that too in the refactoring commit.

parhamsaremi committed:
GitHubCI: refactor

- Merged the restore/build/pack steps for GTK and Xamarin.Forms together.
- Simplified the pack command for Xamarin.Forms.Platform.GTK.

knocte commented:
> Nit: in the first two commits you use lowercase after ':', let's do the same above ^

This nit happened again :)

knocte commented:
```
- merged the restore/build/pack steps for GTK and Xamarin.Forms together.
- simplified the pack command for Xamarin.Forms.Platform.GTK
```
- Please use normal grammatical/ortographical rules for English language in commit messages; so in this case: please uppercase the first letter of the first word in a sentence.
- Missing dot at the end of second sentence.

knocte commented on .github/workflows/main.yml:
```diff
@@ -19,53 +21,23 @@ jobs:
-    
-    - name: Build
-      run: msbuild .\Xamarin.Forms.sln -p:Configuration=Release
+    - name: Restore & Build
```
@parhamsaremi why are you merging the restore and build steps? I never asked for this

parhamsaremi replied:
Afshin used merged steps. So I figured I should do the same.

knocte replied:
Steps should be as small as possible, so that when they fail, you know exactly what thing failed. If Afshin had fat steps, it's my fault for not pointing out in the review.

parhamsaremi replied:
Ummm sure, I'll apply your comments now.

parhamsaremi commented:
> > Nit: in the first two commits you use lowercase after ':', let's do the same above ^
> 
> This nit happened again :)

Oops

knocte commented on .github/workflows/main.yml:
```diff
@@ -20,52 +22,25 @@ jobs:
-      run: msbuild .\Xamarin.Forms.sln -p:Configuration=Release
+      run: |
+        msbuild .\Xamarin.Forms.sln -p:Configuration=${{env.CONFIGURATION}}
+        msbuild Xamarin.Forms.Platform.GTK\Xamarin.Forms.Platform.GTK.csproj /property:Configuration=${{env.CONFIGURATION}}
```
are you sure that the first msbuild command doesn't already build the GTK project? why dont you test removing the second msbuild command to see if it's needed?

knocte commented on .github/workflows/main.yml:
```diff
@@ -20,52 +22,25 @@ jobs:
-    
+      run: |
+        msbuild /restore Xamarin.Forms.sln
+        nuget restore Xamarin.Forms.Platform.GTK\Xamarin.Forms.Platform.GTK.csproj
```
same thing here about the second restore command

knocte commented on .github/workflows/main.yml:
```diff
@@ -20,52 +22,25 @@ jobs:
-
-        nuget pack .nuspec/Xamarin.Forms.nuspec -p "version=$env:DOTNETFORMS_VERSION;Configuration=Release"
+        nuget pack .nuspec/Xamarin.Forms.nuspec -p "version=$version;Configuration=$env:CONFIGURATION"
+        nuget pack './.nuspec/Xamarin.Forms.Platform.GTK.nuspec' -p "version=$version;Configuration=$env:CONFIGURATION;core_version=$version"
```
why do you use single-quotes here while in the first nuget pack command they were not used? let's be consistent

knocte commented:
I see the build is in green in your branch so this seems almost ready to go in!

The only question remaining here, before I hit the "Merge" button is: if your branch is 5.0.0 in your fork, why don't we see more versions of packages being published in https://www.nuget.org/packages/DotNetForms? And if I already included you as owner in https://www.nuget.org/packages/DotNetForms.Platform.GTK, why the last package is 17 days ago and not today?

knocte commented on .github/workflows/main.yml:
```diff
@@ -20,52 +22,23 @@ jobs:
-    
+      run: |
+        msbuild /restore Xamarin.Forms.sln
+
```
nit: looks like this 3-line hunk above is not needed?

parhamsaremi replied:
:+1:

knocte replied:
if you have addressed this, then click on "Resolve conversation" please

knocte commented:
@parhamsaremi you need to add me as owner in https://www.nuget.org/packages/DotNetForms before I can merge this.
